# Introduction to Python -- Class 1

(Date: Feb. 2025)

Notebook content:

1. Hello World!
2. Variables and simple data types
3. Readable printing 
4. Loops
5. Boolean logic


##  Hello World!

Just the typical "Hello Word!" to make sure everything is working. And discussing how to use comments in Python.

In [1]:
print("Hello World!")

Hello World!


In [2]:
# This is a simple, single-line comment in Python!
# print("Hellow World!")
print("Hello!")

Hello!


In [3]:
"""
If you need to make a larger comment section, you would do it like this.
print("Hello World!")
"""
print("Hello!")

Hello!


---

## Variables and simple data types

In this first part of the class, you’ll learn about the different kinds of data you can work with in your Python programs. You’ll also learn how to store your data in variables and how to use those variables in your programs.

Every variable has a type. Here, we will discuss strings, integers, floats, and Booleans. We will learn more about other data types in future classes.

### Strings

Let us replicate the `Hello World` example from above using a variable.

To assign a **value** to a **variable**, we use the `=` sign:

In [4]:
message = "Hello Lovely World!"
print(message)

Hello Lovely World!


### Numbers: integers and floats

Using variables is good programming practice. For instance, suppose someone asks you to to calculate how much money you will get out of a bank deposit with an initial contribution of 100€ and after 7 years at an interest rate of 5%. You could easily calculate it as follows:
$$
\text{Final amount} = \text{Initial amount} * \left( 1 + \frac{\text{interest rate}}{100} \right)^{\text{years}}\,.
$$

In [5]:
100 * (1 + 5 / 100)**7

140.71004226562505

Therefore, we just found out that the answer is 140.71€. 

Instead of doing this, we can define four variables:
* The initial amount of money;
* The interest rate;
* The number of years;
* The final amount of money.

In [6]:
initial = 100
ir = 5
years = 7
final = initial*(1+ir/100)**years
print(final)

140.71004226562505


Evidently, we obtain exactly the same result of 140.71€ but now, if someone changes one of the input data we can simply change the assignment value.

Again, note that the `=` sign is used to **assign a value to a variable**! 

If instead we want to check whether the variable has a given value, then we need to use the `==` sign:

In [7]:
initial == 100

True

In [8]:
initial == 200

False

In [9]:
print(initial)
initial = 200
print(initial)

100
200


In [10]:
initial = 100  # Let us put the value back into what it was for the example.

Every variable has a type, and we can check their type in the following manner:

In [11]:
name = "My name is Filipa"
a = 2
b = 2.
c = "2"
d = True

print(type(name))
print(type(a))
print(type(b))
print(type(c))
print(type(d))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


Python calls any number with a decimal point a **float**. This term is used in most programming languages, and it refers to the fact that a decimal point
can appear at any position in a number.

Essentially, this means that *floats*  are real numbers that computers store with a certain numerical precision. In the case of Python, this precision is $10^{-16}$. You should alywas keep this in mind when writing your programms. Let us see an example

In [12]:
v1 = 1 / 51 * 51
v2 = 1 / 49 * 49

In [13]:
print(v1)
print(v2)

1.0
0.9999999999999999


Did you notice the rounding error?

In [14]:
print(v1 == 1)
print(v2 == 1)

True
False


For this reason, in Python, if you want to test if two floats are equal, you should always do it up to some **precision** smaller than $10^{-16}$. This is how you could do it:

In [15]:
tol = 1e-10
print(abs(v1-v2) < tol)

True


---

## Readable printing

We will now discuss a few features of the printing method that can make printing more pleasant.

In [16]:
pi = 3.1415926535
print(f"({pi:.3f}, {pi:.4})")

(3.142, 3.142)


The example above presents various tricks to make printing more readable:
* The command `f` tells Python that we want to format the string being printed. The curly brakets tell Python to insert the value of the variable inside. 
* The command `pi:.4f` tells Python to print the value of the variable `pi` with 4 decimal places.
* The command `pi:.4` tells Python to print the value of the variable `pi` with 4 digits.

This gives us a great way of printing the outcome of our interest rate exercise:

In [17]:
print(f"After {years} years, an initial deposit of {initial}€ will have grown into {final:.2f}€.")

After 7 years, an initial deposit of 100€ will have grown into 140.71€.


If we are working with large numbers, it could be useful to force Python to use scientific notation. This can be done with `:g`.

Here is an example:

In [18]:
large = 1000000000
print(large)
print(f"A thousand millions: {large:g}.")

1000000000
A thousand millions: 1e+09.


---

## Loops

In many cases we need to perform an action multiple times, or do different things depending on different situations. Instead of doing this one by one, loops help us program such cases quickly and efficiently.

Let us use again our example of the interest rate:

In [19]:
initial = 100
ir = 5
years = 7
final = initial * (1 + ir / 100)**years
print(final)

140.71004226562505


### For loops

Suppose that instead of just wanting to know how much our investment is worth after 7 years, we want to know how much it is worth *every year* for a period of 10 years.

We can write this using a `for` loop:
```
for element in list:           
    <statement_1>              
    <statement_2>              
    ...                        
                               
A new statement after the loop!
```

**Be very careful with the indentation!**

Let us see how this would look like in the case of our example.

In [20]:
for i in range(11):
    final = initial * (1 + ir / 100)**i
    print(f"After {i} years, we would have {final:.2f}€ in the bank.")

After 0 years, we would have 100.00€ in the bank.
After 1 years, we would have 105.00€ in the bank.
After 2 years, we would have 110.25€ in the bank.
After 3 years, we would have 115.76€ in the bank.
After 4 years, we would have 121.55€ in the bank.
After 5 years, we would have 127.63€ in the bank.
After 6 years, we would have 134.01€ in the bank.
After 7 years, we would have 140.71€ in the bank.
After 8 years, we would have 147.75€ in the bank.
After 9 years, we would have 155.13€ in the bank.
After 10 years, we would have 162.89€ in the bank.


There are several important things to be understood from the example above:

1. `i` is our iterator variable (corresponding to the number of years).

2. This variable must have an initial value (here 0) and a final value (here **10**).

3. The command `range(N)` creates a `list` of integers from 0 to N-1 (incremented by 1).<br>
    Alternatively, one can create a list starting in `a` and ending in `b`, with a step of `s` using `range(a, b+1, s)`.

4. The colon tells Python where the conditions for the loop stop and where the commands to be executed begin.

5. The commands to be executed **always** start with an **indentation**.

We can put loops inside of loops as follows:

In [21]:
for m in range(3):
    for k in range(7):
        print(f"La multiplicación de {m} y {k} es igual {m*k}.")

La multiplicación de 0 y 0 es igual 0.
La multiplicación de 0 y 1 es igual 0.
La multiplicación de 0 y 2 es igual 0.
La multiplicación de 0 y 3 es igual 0.
La multiplicación de 0 y 4 es igual 0.
La multiplicación de 0 y 5 es igual 0.
La multiplicación de 0 y 6 es igual 0.
La multiplicación de 1 y 0 es igual 0.
La multiplicación de 1 y 1 es igual 1.
La multiplicación de 1 y 2 es igual 2.
La multiplicación de 1 y 3 es igual 3.
La multiplicación de 1 y 4 es igual 4.
La multiplicación de 1 y 5 es igual 5.
La multiplicación de 1 y 6 es igual 6.
La multiplicación de 2 y 0 es igual 0.
La multiplicación de 2 y 1 es igual 2.
La multiplicación de 2 y 2 es igual 4.
La multiplicación de 2 y 3 es igual 6.
La multiplicación de 2 y 4 es igual 8.
La multiplicación de 2 y 5 es igual 10.
La multiplicación de 2 y 6 es igual 12.


Again. Notice the importance of the indentation.

**Exercise:** Write a loop that calculates the sum of all integers from 1 to N and print out the *final* result.

In [22]:
# Answer 1 is hidden:
N = 10
sum = 0
for i in range(1, N+1):
    sum = sum + i
print(sum)

55


In [23]:
# Answer 2 is hidden:

N = 10
sum = 0
for i in range(1, N + 1):
    sum += i
print(sum)

55


### While loops

A `while` loop has the following form:
```
while condition:               
    <statement_1>              
    <statement_2>              
    ...                        
                               
A new statement after the loop!
```

Important considerations concerning the `while` loop:

1. The `condition` is always a Boolean variable.
2. One of the statements (usually last one) should update the element that is controlling the `condition`.<br>
This is a common mistake that people make (if you forget to update the element controlling the condition, the `while` loop will run forever!)

There are several things that need to be clear before starting a `while` loop:

1. Where or how the loop starts, i.e. what are the initial values ​​of all variables involved.
2. What statements must be repeated within the loop?
3. When the loop ends, i.e. what conditions must become false for the loop to stop.
4. How should variables be updated for each execution of the loop?

Let us have a look at the familiar example of the interest rate!

In [24]:
i = 0  # variable initiation
while i < 11:  # the Boolean condition controlling the loop
    final = initial * (1 + ir / 100)**i
    print(f"After {i} years, we would have {final:.2f}€ in the bank.")
    i += 1  # Updating the element controlling  the condition


After 0 years, we would have 100.00€ in the bank.
After 1 years, we would have 105.00€ in the bank.
After 2 years, we would have 110.25€ in the bank.
After 3 years, we would have 115.76€ in the bank.
After 4 years, we would have 121.55€ in the bank.
After 5 years, we would have 127.63€ in the bank.
After 6 years, we would have 134.01€ in the bank.
After 7 years, we would have 140.71€ in the bank.
After 8 years, we would have 147.75€ in the bank.
After 9 years, we would have 155.13€ in the bank.
After 10 years, we would have 162.89€ in the bank.


**Exercise:** Rewrite the loop for calculating the integers from 1 to N and printing the *final value* now using a `while` loop.

In [25]:
# Answer is hidden:

sum = 0
N = 10
i = 0
while i <= N:
    sum += i
    i += 1
print(sum)

55


**Exercise:** Remember when we used two for loops to calculate the multiplication of the first integers? Can you do the same using a while loop?

In [26]:
# Answer is hidden:

m = 0
while m < 3:
    k = 0
    while k < 7:
        print(f"La multiplicación de {m} y {k} es igual {m*k}.")
        k += 1
    m += 1

La multiplicación de 0 y 0 es igual 0.
La multiplicación de 0 y 1 es igual 0.
La multiplicación de 0 y 2 es igual 0.
La multiplicación de 0 y 3 es igual 0.
La multiplicación de 0 y 4 es igual 0.
La multiplicación de 0 y 5 es igual 0.
La multiplicación de 0 y 6 es igual 0.
La multiplicación de 1 y 0 es igual 0.
La multiplicación de 1 y 1 es igual 1.
La multiplicación de 1 y 2 es igual 2.
La multiplicación de 1 y 3 es igual 3.
La multiplicación de 1 y 4 es igual 4.
La multiplicación de 1 y 5 es igual 5.
La multiplicación de 1 y 6 es igual 6.
La multiplicación de 2 y 0 es igual 0.
La multiplicación de 2 y 1 es igual 2.
La multiplicación de 2 y 2 es igual 4.
La multiplicación de 2 y 3 es igual 6.
La multiplicación de 2 y 4 es igual 8.
La multiplicación de 2 y 5 es igual 10.
La multiplicación de 2 y 6 es igual 12.


### For loops vs while loops

**Comparison between `for` loops and `while` loops**

* In a `for` loop, we need lists. The element controlling the loop iterates over the elements in the list.
* In a `while` loop, we need Boolean conditions that control when the loop terminates. We also need to update the condition during the execution of the loop.


Una diferencia importante entre un bucle for y while es tan fácil debido a la naturaleza booleana de las variables de control en el bucle while.

## Boolean logic

We have seen that an important difference between loops `for` and `while` is that in the latter, the loop is controlled by a Boolean condition. 

This motivates a brief introduction to Boolean logic.

The basic operators of Boolean algebra are: **conjunction** (AND), **disjunction** (OR), and **negation** (NOT). 

Denoting **True** as 1 and **False** as 0, the *truth tables* of these operators are:

<center>

| **a** | **b** | **a AND b** |
|:-:|:-:|:-:|
| 0 | 0 | 0 | 
| 0 | 1 | 0 |  
| 1 | 0 | 0 |
| 1 | 1 | 1 |

<br>

| **a** | **b** | **a OR b** |
|:-:|:-:|:-:|
| 0 | 0 | 0 | 
| 0 | 1 | 1 |  
| 1 | 0 | 1 |
| 1 | 1 | 1 |

<br>

| **a** | **NOT a** | 
|:-:|:-:|
| 0 | 1 | 
| 1 | 0 |   
</center>



**Exercise**<br>

Let $a=0$, $b=1.2$.
What are the values ​​of the following boolean operations?<br>

1. $a \geq 0$ AND $b < 1$. 
2. $a \geq 0$ OR $b < 1$.
3. $a > 0$ OR $b > 1$.
4. $a > 0$ OR NOT $b > 1$.
5. $-1 < a  \leq 0$ (same as $-1 < a$ AND $ a \leq 0$).
6. NOT ($a > 0$ OR b > 0).

In [27]:
a = 0
b = 1.2
print(a >= 0 and b < 1)
print(a >= 0 or b < 1)
print(a > 0 or b > 1)
print(a > 0 or not b > 1)
print(-1 < a <= 0)
print(not (a > 0 or b > 0))

False
True
True
False
True
False


We can use Boolean logic to make even more refined loops.

Let us return to our example with the bank interest rate.

Suppose we want to take our money from the bank either after 20 years, or when our money doubles. How can we determine which of the conditions happens first and when we can take the money out?

In [28]:
# Answer is hidden

initial = 100
ir = 5
years = 0
x = 0
while (years <= 20 and x <= 2.):
    final = initial * (1 + ir / 100)**years
    x = final / initial
    print(years, final)
    years += 1
print(f"If we remove the money after {years-1} years, it will be worth {final}€.")

0 100.0
1 105.0
2 110.25
3 115.76250000000002
4 121.55062500000003
5 127.62815625000003
6 134.00956406250003
7 140.71004226562505
8 147.7455443789063
9 155.13282159785163
10 162.8894626777442
11 171.0339358116314
12 179.585632602213
13 188.56491423232367
14 197.99315994393987
15 207.89281794113688
If we remove the money after 15 years, it will be worth 207.89281794113688€.


---